lib/helpers/common.js

```diff
@@ -67,7 +69,22 @@ function flatten(update, path, options, schema) {
  * ignore
  */
 
-function modifiedPaths(update, path, result) {
+function modifiedPaths(update, path, result, recursion = null) {
+  if (recursion == null) {
+    recursion = {
+      raw: { update, path },
+      trace: new WeakSet()
+    };
+  }
+
+  if (recursion.trace.has(update)) {
+    throw new MongooseError(`a circular reference in the update value, updateValue:
+${util.inspect(recursion.raw.update, { showHidden: false, depth: 1 })}
+updatePath: ${recursion.raw.path}
+-------`);
```

Comment:
Minor nitpick: this `----` is a bit messy. But I'll clean that up after merging, no need for that to be a blocker.